### ISSUE TITLE
Call updateTimestamp() before taking kernel lock for IRQs?

### ISSUE THREAD

**Issue by Indanz** [seL4/seL4#844]
System calls are initiated by the currently running task, so charging them makes sense. The same is not true for interrupts, especially not the scheduling timer interrupt causing a context switch.

It may seem it does not matter whether a task is charged for the context switch at the beginning of its period or at the end, but it does matter for periodic tasks that call seL4_yield(): Counter intuitively, such tasks run less often if they are not charged at the beginning. It is because they will run every period + kernel lock take time (~= WCET), instead of every period.

This introduces an accumulating error between the system clock and timers managed by such tasks. Practically it means any timers handled by such task will sometimes trigger one period later than expected, if they use absolute time based on the system clock.

On the system I use the error is about 1 ms per four seconds. It is a problem because I have a polling driver that wants to poll every millisecond, now it's sometimes triggers twice after two milliseconds, introducing unnecessary jitter into the system.

In my case there are ways of working around this issue, but the fundamental question is whether such drift for periodic tasks is unwanted or not.

Below a patch that eliminates drift for periodic tasks on my system (Aarch64). I also need the -1 in setDeadline(), but sadly getTimerPrecision() uses usToTicks(TIMER_PRECISION), with TIMER_PRECISION in microseconds instead of ticks.
```
diff --git a/src/api/syscall.c b/src/api/syscall.c
index 40f6f0690280..9155503de9dd 100644
--- a/src/api/syscall.c
+++ b/src/api/syscall.c
@@ -41,7 +41,6 @@ exception_t handleInterruptEntry(void)
     irq = getActiveIRQ();
 #ifdef CONFIG_KERNEL_MCS
     if (SMP_TERNARY(clh_is_self_in_queue(), 1)) {
-        updateTimestamp();
         checkBudget();
     }
 #endif
diff --git a/src/arch/arm/c_traps.c b/src/arch/arm/c_traps.c
index 740d00313d9a..550bf78bbaf4 100644
--- a/src/arch/arm/c_traps.c
+++ b/src/arch/arm/c_traps.c
@@ -89,6 +89,7 @@ void VISIBLE NORETURN c_handle_instruction_fault(void)

 void VISIBLE NORETURN c_handle_interrupt(void)
 {
+    updateTimestamp();
     NODE_LOCK_IRQ_IF(IRQT_TO_IRQ(getActiveIRQ()) != irq_remote_call_ipi);
     c_entry_hook();

diff --git a/src/kernel/thread.c b/src/kernel/thread.c
index 7736ce6e01ef..fe2786f79161 100644
--- a/src/kernel/thread.c
+++ b/src/kernel/thread.c
@@ -585,7 +585,7 @@ void setNextInterrupt(void)
         next_interrupt = MIN(refill_head(NODE_STATE(ksReleaseHead)->tcbSchedContext)->rTime, next_interrupt);
     }

-    setDeadline(next_interrupt - getTimerPrecision());
+    setDeadline(next_interrupt - getTimerPrecision() - 1);
 }

 void chargeBudget(ticks_t consumed, bool_t canTimeoutFault)
```

**Comment by Indanz** [comment]
Thinking about it a bit more, I think the current behaviour is unwanted because it makes it possible for low priority tasks to postpone execution of higher priority periodic tasks that use their full budget (either by using it or by yielding).

It also makes it harder to design systems if multiple periodic tasks have different amounts of drift and get out of sync with each other. It makes the system less deterministic.

**Comment by kent-mcleod** [comment]
> I also need the -1 in setDeadline(), but sadly getTimerPrecision() uses usToTicks(TIMER_PRECISION), with TIMER_PRECISION in microseconds instead of ticks.

This is unrelated to moving the position of updateTimestamp right?

**Comment by Indanz** [comment]
> This is unrelated to moving the position of updateTimestamp right?

Correct. The -1 in setDeadline is only needed to make the actual period exactly match the configured period. Without this behaviour is still deterministic. The -1 can easily be fixed with introducing a new config option, because it depends on the platform.

Your understanding is correct. On a quad core system, worst case time to take the lock is 3 times the single core WCET.

> The intention of the MCS design is that updateTimestamp is performed as early in the kernel operation instruction stream as possible

In that case updateTimestamp() should always be called before taking the lock, even for the fast path. My understanding is that it only matters when/if updateTimestamp() is called in case of reschedule, and then it needs to be called at the time you want the previous task's period to end and the new one to start. But if the proofs do assume it's always taken at the very beginning it sooner or later will run into issues because the implementation is different.

> One potential issue is whether updateTimestamp requires the lock to be already acquired in order to update kernel state.

It only touches node data, so it should be fine. It's actually very hard for an implementation to not be thread safe, because even if it uses global data, it would be in a read-only matter. However, we should double check all getCurrentTime() implementations. As far as I can tell they are all fine, including all the implementations in include/drivers/timer/. Current updateTimestamp is:

```
static inline void updateTimestamp(void)
{
    time_t prev = NODE_STATE(ksCurTime);
    NODE_STATE(ksCurTime) = getCurrentTime();
    assert(NODE_STATE(ksCurTime) < MAX_RELEASE_TIME);
    time_t consumed = (NODE_STATE(ksCurTime) - prev);
    NODE_STATE(ksConsumed) += consumed;
    if (numDomains > 1) {
        if ((consumed + MIN_BUDGET) >= ksDomainTime) {
            ksDomainTime = 0;
        } else {
            ksDomainTime -= consumed;
        }
    }

}
```

**Comment by Indanz** [comment]
Hm, is ksDomainTime supposed to be per node or system-wide? I would expect it to be node state, not global state.

Is MCS + SMP + numDomains > 1 supported?

**Comment by kent-mcleod** [comment]
> Is MCS + SMP + numDomains > 1 supported?

SMP + numDomans > 1 is not supported: 
https://github.com/seL4/seL4/blob/master/config.cmake#L292-L297

**Comment by kent-mcleod** [comment]
> It only touches node data, so it should be fine.

There are some entries in node data that are updated by other cores (search for usages of `NODE_STATE_ON_CORE`), but it appears that ksConsumed and ksCurTime are only ever accessed by the local core.  So I still agree that it doesn't create a race to move updateTimestamp outside of the lock given that numDomains == 1 when there is a lock.

We would only need to check the getCurrentTime() implementations for platforms that have multicore.

**Comment by Indanz** [comment]
ksCurTime may be read cross-core. ksConsumed is never accessed via `NODE_STATE_ON_CORE`. If it was we had to check whether both reads had to be atomic in regard to each other.

refill_new(), refill_ready(), refill_update() and refill_unblock_check() all check ksCurTime via `NODE_STATE_ON_CORE`.

- refill_new: Safe. It sets rTime to ksCurTime.
- refill_ready: Seems safe.
- refill_update: No idea why it changes rTime there. I think it's safe to remove that assignment. But it calls remoteTCBStall if sc->scTcb is set, so remote core is blocked anyway.
- refill_unblock_check: Same as refill_update, except that it also sets it's own ksReprogram to true.

**Comment by kent-mcleod** [comment]
> refill_unblock_check: No idea why it changes rTime there. I think it's safe to remove that assignment.

This is necessary to prevent accumulation of budgets over long idle times. A sporadic SC will lose any accumulated budget whenever it becomes unblocked.  Based on the value of `seL4_SchedContextFlag` SCs will have refill_unblock_check called at different times which affect what blocking conditions constitute losing accumulated budget.

> refill_new(), refill_ready(), refill_update() and refill_unblock_check() all check ksCurTime via NODE_STATE_ON_CORE.

I agree that they all appear safe too.

There's probably a different issue with these code paths anyway.  The ksCurTime value is supposed to be a cache of getCurrentTime(). But for remote cores this can be very stale as it is only updated on the most recent kernel entry of its core.
For example with the refill_unblock_check() case, a remote SC could become unblocked and get a rTime that is for a very old ksCurTime which would mean it potentially loses most or all of its budget when it is scheduled by it's own core.

**Comment by Indanz** [comment]
> There's probably a different issue with these code paths anyway. The ksCurTime value is supposed to be a cache of getCurrentTime(). But for remote cores this can be very stale as it is only updated on the most recent kernel entry of its core.

This was my conclusion as well.

> This is necessary to prevent accumulation of budgets over long idle times. A sporadic SC will lose any accumulated budget whenever it becomes unblocked.

`refill_unblock_check()` adds overlapping budgets (rAmounts) together, accumulated budget is made available from now on. What it does lose are past rTime refill moments and in that sense it loses budget, as it can't catch up with refills. I thought this was only done when actually scheduling a task, but it's done more often for sporadic tasks, also cross-core.

> a remote SC could become unblocked and get a rTime that is for a very old ksCurTime which would mean it potentially loses most or all of its budget when it is scheduled by it's own core.

Isn't the problem the opposite, that it doesn't lose accumulated budget? (Or gains it equal to the difference with the next ksCurTime, in case of refill_new?)

All in all it seems more correct to use own ksCurTime than the SC's node one. But some architectures (e.g. older x86) may have async system locks. If it's rare enough I think it's fine to say seL4 requires synchronous system clocks, it may make verification easier too if there is a global time.

At least after my change a remote node's ksCurTime is updated after an ipi, making the problem less severe.

**Comment by Indanz** [comment]
I could add a test that checks for how synchronous periodic tasks are with the system clock, but then TIMER_OVERHEAD_TICKS needs to be set correctly for all tested platforms.

**Comment by Indanz** [comment]
Riscv is okay: "The real-time clocks of all hardware threads in a single user application should be synchronized to within one tick of the real-time clock."

x86 okay if not ancient.

aarch64 is okay.

arm32: arm_global.h is okay. Exynos timer seems okay. am335x and omap3430 are single-core.

All-in-all I think those `NODE_STATE_ON_CORE(ksCurTime)` should be replaced with `NODE_STATE(ksCurTime)` calls.

**Comment by kent-mcleod** [comment]
> All-in-all I think those `NODE_STATE_ON_CORE(ksCurTime)` should be replaced with `NODE_STATE(ksCurTime)` calls.

This should probably become it's own issue. I agree this is reasonable given that the SMP configuration is targeting SMP systems where it's reasonable to expect clock synchronization.

**Comment by kent-mcleod** [comment]
> Isn't the problem the opposite, that it doesn't lose accumulated budget? (Or gains it equal to the difference with the next ksCurTime, in case of refill_new?)

I may have gotten this backwards.  I was worrying that `ksConsumed` would get out of sync for the target core because `rTime` would be old.  But you are right that it will instead lead to some bandwidth amplification because the rTime could be further in the past than it's supposed to be

**Comment by kent-mcleod** [comment]
> System calls are initiated by the currently running task, so charging them makes sense. The same is not true for interrupts, especially not the scheduling timer interrupt causing a context switch.

I think it would be better for the updateTimestamp() to be moved outside of the lock for all kernel entries instead of just for IRQs.

**Comment by Indanz** [comment]
> I think it would be better for the updateTimestamp() to be moved outside of the lock for all kernel entries instead of just for IRQs. Thoughts?

I tried that, and that didn't help with the issue I observed and the diff was much larger, so I dropped it. I can redo it if you like. It involves removing `updateTimestamp()` from `MCS_DO_IF_BUDGET()`. It would be more consistent to do it always. Some VCPU traps may be more like irqs too.

Conceptually it makes sense to me that the task causing a cost pays for it, and in the case of doing a system call a large part of that cost can be waiting for the lock, but this is only part of the cost. If attributing the whole system call cost to the calling taks, the updateTimestamp() should be called just before context switching. That is an independent issue though.

The best approach depends on what MCS time model is being implemented, I think things like this should be verification guided.

**Comment by Indanz** [comment]
@kent-mcleod I updated #847 to always call `updateTimestamp()` first.

@axel-h I disagree that this is an enhancement. Considering this fixes time accounting and that it prevents low-priority tasks from delaying high-priority tasks, I would say it is a bugfix.

**Comment by axel-h** [comment]
@Indanz I do not have a preference here, can mark it as bug also if there is no objection.

**Comment by axel-h** [comment]
I wonder if this change is worth an entry in the `CHANGES` file also, especially if it considered a bugfix. Could to add this in the PR?

**Comment by Indanz** [comment]
I do not know what the policy is for `CHANGES`. It's only a bugfix on SMP, which is not verified yet. I expect a lot more bugfixes will pop-up if SMP verification is started. It's also a minor bug which probably doesn't impact users much (otherwise people would have noticed it before). If there is no formal time model for MCS yet, then this whole point is moot, as it's undefined what the "correct" behaviour is.

**Comment by kent-mcleod** [comment]
> The best approach depends on what MCS time model is being implemented, I think things like this should be verification guided.

Yea, we also still need to get to get input from someone working on the MCS verification that they are prepared to accept this change. It's likely that for unicore we need to keep `updateTimestamp()` in the same place and only move it for SMP. 

> I tried that, and that didn't help with the issue I observed

Do you mean moving updateTimestamp() out everywhere didn't help? Isn't that what is being proposed now?

**Comment by kent-mcleod** [comment]
> On the system I use the error is about 1 ms per four seconds. It is a problem because I have a polling driver that wants to poll every millisecond, now it's sometimes triggers twice after two milliseconds, introducing unnecessary jitter into the system.

Are you able to expand on how your system is setup?
You have an SC with a configured budget and period? And also user level timers that are generating start events or are you relying on the refresh of the period to wake up the thread?

**Comment by kent-mcleod** [comment]
Also is the SC configured with the Sporadic or ConstantBandwidth refill policy?

**Comment by pingerino** [comment]
> This introduces an accumulating error between the system clock and timers managed by such tasks. Practically it means any timers handled by such task will sometimes trigger one period later than expected, if they use absolute time based on the system clock.
 
I had always thought we should allow access to the kernel timer for user level - e.g the kernel provides a virtual async endpoint which you can ask for irqs on at specific timeouts. This would leverage the existing release queue and remove these kinds of timing issues. Didn't ever get far enough to implement it though.

Without this, using the kernel to control time based scheduling is not that precise - the time budgets on scheduling contexts only work if you have a very detailed and accurate analysis of your system and potential blocking times, and that's on a single core. For multicore the lock makes this more complicated. An ideal design would  pull much more logic out of the kernel lock to make a real partitioned scheduler.

> System calls are initiated by the currently running task, so charging them makes sense. The same is not true for interrupts, especially not the scheduling timer interrupt causing a context switch.

When it comes to multicore + realtime, any blocking time has to be factored into your budget and schedule and has to be billed to the blocked thread (or you'll miss deadlines).

**Comment by Indanz** [comment]
> It's likely that for unicore we need to keep updateTimestamp() in the same place and only move it for SMP.

That seems unlikely to me, at least in the long term. Short term it avoids proof updates I suppose, but I expect those to be a matter of moving bits of proofs around, but not really changing.

> Do you mean moving updateTimestamp() out everywhere didn't help? Isn't that what is being proposed now?

Sorry for the confusing phrasing. I meant that to solve the particular issue I ran into, only moving the checkBudget calls for irqs was enough.

> You have an SC with a configured budget and period? And also user level timers that are generating start events or are you relying on the refresh of the period to wake up the thread?

The timer task has a period of 1 ms and a budget of 100 us, with ConstantBandwidth refill policy (seL4_SchedContext_NoFlag).

Tasks that want to wake up at specific time tell the absolute time they want to be woken up. One polling task wants to wake up every millisecond, and after it was woken up it increments its requested wake-uptime with 1 ms. They are configured as round-robin.

At the start of each period the timer task reads the current time (via `cntpct`) and checks which tasks want to wake up this period and sends them a signal. Then it calls seL4_Yield() to wait for the next period.

Just to be clear, some jitter caused by scheduling jitter and rounding errors is to be expected, but long-term I would expect that the timer task would run at average thousand times per second and that it would wake up the polling task every millisecond, not occasionally twice after 2 ms because of drift in the timer task's period.

> I had always thought we should allow access to the kernel timer for user level - e.g the kernel provides a virtual async endpoint which you can ask for irqs on at specific timeouts. This would leverage the existing release queue and remove these kinds of timing issues. Didn't ever get far enough to implement it though.

If there was something like this I wouldn't need my timer task at all. It requires a system call to update that wake-up time though, lower overhead would be to have an option to pass the wake-up time to seL4_Wait/seL4_Recv, but that may be harder to implement.

> When it comes to multicore + realtime, any blocking time has to be factored into your budget and schedule and has to be billed to the blocked thread (or you'll miss deadlines).

Say there are N tasks with each the same period, but different budgets and all budgets add up to the period, would you expect this to be schedulable and all tasks having access to their full budget every period?

I'll try to create a simple test to demonstrate the problem.

**Comment by pingerino** [comment]
> Say there are N tasks with each the same period, but different budgets and all budgets add up to the period, would you expect this to be schedulable and all tasks having access to their full budget every period?

If they are all the same priority and never get preempted, sure. Otherwise we need to look at the rate monotonic analysis, and the impacts of the sporadic server on the budget consumption if there are preemptions.

**Comment by kent-mcleod** [comment]
> The timer task has a period of 1 ms and a budget of 100 us, with ConstantBandwidth refill policy (seL4_SchedContext_NoFlag).

The ConstantBandwidth policy would also be contributing to the drift you are experiencing. It's applies a sliding window constraint to refill replenishments where the next replenishment is scheduled 1 period in the future from the last time the thread was activated. So if there's a higher priority thread running when your thread becomes replenished it will be delayed until the higher priority thread finishes and also it's future periodic replenishments will be delayed by the same amount.  Without this delay, a thread could encounter a "back to back" scenario where for it's first period it's delayed until right before the period is due to replenish it could run continuously for 2 budgets worth of time as the first budget would replenish immediately after it was consumed. 

The Sporadic policy instead allows the back to back scenario where the delayed start time of the thread doesn't cause it's next replenishment to be delayed. This also allows the budget to accumulate if the thread spends more than a single period preempted by a higher priority thread and allows it to "catch-up" any lost execution time once it is scheduled again. However this requires that the SC attached to the thread doesn't go through any kernel paths that call refill_unblock_check on it.  If it's only ever being preempted and sending signals then refill_unblock_check won't be called on it. 

So under a ConstantBandwidth policy, while moving the `updateTimestamp()` to before the lock in the IRQ path will apparently reduce the delayed start to your thread at the start of its period each time, if it's ever preempted by any other higher priority threads while it's executing it will still drift.  Also any small error due to the timer interrupt coming in late will also accumulate. 

Have you investigated using the Sporadic policy at all?

**Comment by Indanz** [comment]
> Ideally this would be outside the lock too.

I like the idea of lockless node local scheduling.

> So if there's a higher priority thread running when your thread becomes replenished it will be delayed until the higher priority thread finishes and also it's future periodic replenishments will be delayed by the same amount.

There are literally no other tasks running on that core, and even if they were, the timer task has highest priority. And these kind of drifts are expected and ruled out to be the case here. This issue is not about drift in general, but only about drift caused by the period being extended by the lock take time.

> However this requires that the SC attached to the thread doesn't go through any kernel paths that call refill_unblock_check on it.

Well, let's just say that refill_unblock_check will mess up any constant periodicity. The Sporadic policy seems better suited for cases where the goal is to divide and limit total CPU time. For time periodic tasks that want hard deadlines and where there is no point in catching up, the Constant bandwidth policy is more suitable.

> Have you investigated using the Sporadic policy at all?

No, because I didn't think that would made a difference for a task that has the highest priority. In that case the behaviour of Sporadic or Constant bandwidth should be the same, right? (Assuming all the task does is sending signals.)

**Comment by kent-mcleod** [comment]
> No, because I didn't think that would made a difference for a task that has the highest priority. In that case the behaviour of Sporadic or Constant bandwidth should be the same, right? (Assuming all the task does is sending signals.)

In the case where there's a single task at the highest priority calling only seL4_Signal and seL4_Yield, the difference would be that ConstantBandwidth calls refill_unblock_check when switching from the idle task back to the main task.  The effect of this is setting `refill_head(sc)->rTime` to ksCurTime.  Without the unblock check the value of refill_head(sc)->rTime would be left as what it was set to when seL4_Yield was called: `rTime = refill_head(sc)->rTime + sc->scPeriod`.

The period error you are experiencing is because the timer interrupt is programmed for:
```
    if (NODE_STATE(ksReleaseHead) != NULL) {
        next_interrupt = MIN(refill_head(NODE_STATE(ksReleaseHead)->tcbSchedContext)->rTime, next_interrupt);
    }

    setDeadline(next_interrupt - getTimerPrecision());
```
With ConstantBandwidth, when the timer interrupt fires and `updateTimestamp()` is eventually called, if the new value of ksCurTime is greater than the value of rTime used to program the timeout then that difference is essentially added onto the period as you say. With Sporadic, the difference is not added onto the period and the next time setDeadline is called it will only be for rTime + sc->scPeriod which means that the period error doesn't accumulate.

**Comment by kent-mcleod** [comment]
In any case I think we are still in agreement that the updateTimestamp() should be moved out of the lock on SMP to continue following the model that it's called as close to the start of the kernel operation as possible.

**Comment by Indanz** [comment]
> With Sporadic, the difference is not added onto the period and the next time setDeadline is called it will only be for rTime + sc->scPeriod which means that the period error doesn't accumulate.

I can confirm that with `seL4_SchedContext_Sporadic` there does not seem to be any drift (tested for 5 minutes).

**Comment by kent-mcleod** [comment]
I think we now wait for @lsf37 to come back from leave.

@Indanz, a separate question, can I send you an email with a question about the 2022 seL4 summit?

**Comment by Indanz** [comment]
@kent-mcleod, sure, anyone may always e-mail me.
